edi_storage_oca/migrations/16.0.1.0.0/pre-migrate.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2024 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import SUPERUSER_ID, api
+from odoo.tools.sql import column_exists, create_column, rename_column
+
+
+def migrate(cr, version):
+    if not version:
+        return
+    env = api.Environment(cr, SUPERUSER_ID, {})
+
+    required_modules = ("storage_backend", "edi_storage_oca")
+    if not env["ir.module.module"].search(
+        [("name", "in", required_modules), ("state", "=", "installed")]
+    ):
+        return
+    if not column_exists(cr, "edi_backend", "new_storage_id"):
+        create_column(cr, "edi_backend", "new_storage_id", "int4")
+    _create_fs_storage_records(env)
+    if column_exists(cr, "edi_backend", "new_storage_id"):
+        rename_column(cr, "edi_backend", "storage_id", "old_storage_id")
+        rename_column(cr, "edi_backend", "new_storage_id", "storage_id")
+
+
+def _create_fs_storage_records(env):
+    # Create a fs.storage record for each backend.storage
+    storage_backend_records = env["storage.backend"].search([])
+    if not storage_backend_records:
+        return
+
+    # make sure all backend_type can be mapped even if corresponding modules
+    # have not been migrated (on purpose because we should switch to fs_storage)
+    selection = [
+        ("filesystem", "Filesystem"),
+        ("ftp", "FTP"),
+        ("sftp", "SFTP"),
+        ("s3", "S3"),
+    ]
+    env["storage.backend"]._fields["backend_type"].selection = selection
+    env["storage.backend"]._fields["backend_type_env_default"].selection = selection
```

Comment:
Hi Nils, about this one, I don't understand why we need it, could you explain a little bit here for me 😅

Reply:
- here is the full traceback I had when loading a db with a `storage.backend` with `backend_type = 'ftp'` in v16:

```
  File "/odoo_env/src/odoo/odoo/api.py", line 1004, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'storage.backend(1,).backend_type_env_default'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/odoo_env/src/odoo/odoo/service/server.py", line 1310, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-16>", line 2, in new
  File "/odoo_env/src/odoo/odoo/tools/func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "/odoo_env/src/odoo/odoo/modules/registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/odoo_env/src/odoo/odoo/modules/loading.py", line 484, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "/odoo_env/src/odoo/odoo/modules/loading.py", line 372, in load_marked_modules
    loaded, processed = load_module_graph(
  File "/odoo_env/src/odoo/odoo/modules/loading.py", line 183, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/odoo_env/src/openupgrade/openupgrade_framework/odoo_patch/odoo/modules/migration.py", line 18, in migrate_module
    MigrationManager.migrate_module._original_method(self, pkg, stage)
  File "/odoo_env/src/odoo/odoo/modules/migration.py", line 189, in migrate_module
    migrate(self.cr, installed_version)
  File "/odoo_env/src/oca-edi-framework/edi_storage_oca/migrations/16.0.1.0.0/pre-migrate.py", line 22, in migrate
    _create_mapping_records(env)
  File "/odoo_env/src/oca-edi-framework/edi_storage_oca/migrations/16.0.1.0.0/pre-migrate.py", line 39, in _create_mapping_records
    storage_backend_records.filtered(
  File "/odoo_env/src/odoo/odoo/models.py", line 5444, in filtered
    return self.browse([rec.id for rec in self if func(rec)])
  File "/odoo_env/src/odoo/odoo/models.py", line 5444, in <listcomp>
    return self.browse([rec.id for rec in self if func(rec)])
  File "/odoo_env/src/oca-edi-framework/edi_storage_oca/migrations/16.0.1.0.0/pre-migrate.py", line 40, in <lambda>
    lambda record: record.backend_type == "filesystem"
  File "/odoo_env/src/odoo/odoo/fields.py", line 1210, in __get__
    self.compute_value(recs)
  File "/odoo_env/src/odoo/odoo/fields.py", line 1392, in compute_value
    records._compute_field_value(self)
  File "/odoo_env/src/odoo/odoo/models.py", line 4232, in _compute_field_value
    fields.determine(field.compute, self)
  File "/odoo_env/src/odoo/odoo/fields.py", line 98, in determine
    return needle(*args)
  File "/odoo_env/src/oca-server-env/server_environment/models/server_env_mixin.py", line 277, in _compute_server_env
    record._compute_server_env_from_default(field_name, options)
  File "/odoo_env/src/oca-server-env/server_environment/models/server_env_mixin.py", line 261, in _compute_server_env_from_default
    self[field_name] = self[default_field]
  File "/odoo_env/src/odoo/odoo/models.py", line 5948, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File "/odoo_env/src/odoo/odoo/fields.py", line 1210, in __get__
    self.compute_value(recs)
  File "/odoo_env/src/odoo/odoo/fields.py", line 1392, in compute_value
    records._compute_field_value(self)
  File "/odoo_env/src/odoo/odoo/models.py", line 4232, in _compute_field_value
    fields.determine(field.compute, self)
  File "/odoo_env/src/odoo/odoo/fields.py", line 101, in determine
    return needle(records, *args)
  File "/odoo_env/src/odoo/addons/base_sparse_field/models/fields.py", line 53, in _compute_sparse
    record[self.name] = values.get(self.name)
  File "/odoo_env/src/odoo/odoo/models.py", line 5957, in __setitem__
    return self._fields[key].__set__(self, value)
  File "/odoo_env/src/odoo/odoo/fields.py", line 1305, in __set__
    self.write(protected_records, value)
  File "/odoo_env/src/odoo/odoo/fields.py", line 1125, in write
    cache_value = self.convert_to_cache(value, records)
  File "/odoo_env/src/odoo/odoo/fields.py", line 2735, in convert_to_cache
    raise ValueError("Wrong value for %s: %r" % (self, value))
ValueError: Wrong value for storage.backend.backend_type_env_default: 'ftp'
```

- that's because `storage_backend_ftp` was not migrated to v16 (and should not be, the goal is to move to fs_storage): so value `ftp` is not added to the possible values for `backend_type` (see [here](https://github.com/OCA/storage/blob/15.0/storage_backend_ftp/models/storage_backend.py#L10) in v15)
- so we add it ourselves just for the migration of data